bgpv2: Update status of CiliumBGPNodeConfig when gobgp state changes

harsimran-pabla opened the pull request:
This change introduces state tracking of GoBGP. One use case of state tracking is updating the status field of the CiliumBGPNodeConfig resource. 

Change is split into 4 parts 
- [k8s] Updating the status field of CIliumBGPNodeConfig. **This PR is making a CRD status field change.** The `status` field has never been set before, as this is a new implementation, and some iterations were necessary for it. 

- [GoBGP/Manager] Introduce plumbing to get notifications when GoBGP state changes.
- [CRD Status Reconciler] Reconciler, which updates the status field of the CiliumBGPNodeConfig resource.
- [Jitter in status updates] Status update reconciler to update API server at ticker interval with added jitter.

<details>
<summary>Sample output of the status field</summary>

```
 kubectl get ciliumbgpnodeconfigs/bgpv2-cplane-dev-service-worker -o json | jq '.status'
{
  "bgpInstances": [
    {
      "localASN": 65001,
      "name": "65001",
      "peers": [
        {
          "establishedTime": "2024-06-27T17:30:27Z",
          "name": "65000",
          "peerASN": 65000,
          "peerAddress": "fd00:10:0:1::1",
          "peeringState": "established",
          "routeCount": [
            {
              "advertised": 9,
              "afi": "ipv4",
              "received": 8,
              "safi": "unicast"
            },
            {
              "advertised": 2,
              "afi": "ipv6",
              "received": 1,
              "safi": "unicast"
            }
          ],
          "timers": {
            "appliedHoldTimeSeconds": 90,
            "appliedKeepaliveSeconds": 30
          }
        }
      ]
    }
  ]
}
```
</details>

```release-note
bgpv2: update status field of CiliumBGPNodeConfig CRD
```

harsimran-pabla committed:
bgpv2: update status field of CiliumBGPNodeConfig

Updating the CRD status field. Uptime is replaced by EstablishedTime,
which shows when BGP session state went into established.

This change also adds more granular route count per AFI/SAFI and removes
global received and advertised route count.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: track state changes in gobgp and init state reconcilers

Introduce state change tracking in gobgp. On detecting route table or
peer state changes, gobgp will notify the caller via goroutine trigger.
Adding shutdown flag so watcher does not try to write on closed channel.

Introducing state reconciler infra structure as well which go over all
state reconcilers to notify them to reconcile with gobgp instance.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: updating CiliumBGPNodeConfig status fields

Introducing CRD status update reconciler. This reconciler updates local
desired status and another goroutine periodically checks running and
expected statuses. On change it will update status field of
CiliumBGPNodeConfig resource.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -0,0 +1,277 @@
+				peerStatus.LastUpdateTimestamp = ptr.To[string](time.Now().Format(time.RFC3339))
+
+				// Uptime is in nanoseconds, round to seconds
+				peerStatus.Uptime = ptr.To[string](time.Duration(runningPeerState.UptimeNanoseconds).Round(time.Second).String())
```
Exposing uptime in seconds would mean we need to update the status upon each reconcile (every 5s). What about exposing last state change as a timestamp instead?

harsimran-pabla replied:
Yeah, the `uptime` field in the CRD is a bit tricky. In this change, I introduced the `last update timestamp` field so users can confirm `uptime` and `last update timestamp` to conclude how long this session is up. The tricky bit is that the CRD is updated only when a state change is detected. 

One more option: We should have a low cadence loop, say, every 30 seconds, where we get state from gobgp. (This is all done via reconciler.) This way, we keep uptime somewhat accurate. In case a state change happens between the 30-second interval, we will update the CRD within 5 seconds via the normal Reconcile trigger.

rastislavs replied:
I wouldn't be concerned about the uptime not being accurate too much, but what I mainly meant was that this comparison:
```
r.desiredStatus.DeepEqual(r.runningStatus)
```
will always result as non-equal if we keep a time in seconds (that keeps counting every second) in the status. If it was a timestamp of the last transition instead, the value does not change until the peering state changes.

harsimran-pabla replied:
I am not following this concern correctly. 

DesiredStatus is only updated when Reconcile is called, which happens whenever there is a state change ( peer or routing table ). Uptime in `desired status` is only set during this Reconcile. Once `running status` matches it and the BGP session stabilizes, we will not have a change to `uptime` in `desired status.`

harsimran-pabla replied:
I see concerns that each state reconcile trigger; we will also update CRD since the `uptime` will change.

harsimran-pabla replied:
Hmm, this would require making non-backwards compatible change to CRD, since uptime is already part of the status field.

harsimran-pabla replied:
> I see concerns that each state reconcile trigger; we will also update CRD since the uptime will change.

Is this a big deal, though? It just keeps the status field up-to-date.

rastislavs replied:
Ah, sorry I missed that DesiredStatus is only updated when Reconcile is called, so that's really not an issue.

I still think that exposing something like `EstablishedTime` (timestamp) instead of `Uptime` (duration) + `LastUpdateTimestamp` would be better. But also have nothing against keeping it in the original version if it does not cause more frequent status updates. 

Not sure if low cadence periodic update from gobgp is that necessary, I probably would not bother with that - but up to you.

nit: please note the [naming conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-conventions) rule:
> The name of a field that specifies the time at which something occurs should be called somethingTime. Do not use stamp (e.g., creationTimestamp).

rastislavs replied:
Ah good point regarding non-compatible CRD change, but as status was never populated before, I think a change here it could be allowed?

harsimran-pabla replied:
> as status was never populated before, I think a change here it could be allowed?

Yes, that's a possibility. I will update CRD in that case. 

Couple of changes
- Remove `uptime` and `last update timestamp`,  and introduce `established time.`
- Remove sum of route counts, only keep per afi/safi received and advertised route count.

harsimran-pabla replied:
With the latest change, established time is published.

Sample output 

```
 kubectl get ciliumbgpnodeconfigs/bgpv2-cplane-dev-service-worker -o json | jq '.status'
{
  "bgpInstances": [
    {
      "localASN": 65001,
      "name": "65001",
      "peers": [
        {
          "establishedTime": "2024-06-27T17:30:27Z",
          "name": "65000",
          "peerASN": 65000,
          "peerAddress": "fd00:10:0:1::1",
          "peeringState": "established",
          "routeCount": [
            {
              "advertised": 9,
              "afi": "ipv4",
              "received": 8,
              "safi": "unicast"
            },
            {
              "advertised": 2,
              "afi": "ipv6",
              "received": 1,
              "safi": "unicast"
            }
          ],
          "timers": {
            "appliedHoldTimeSeconds": 90,
            "appliedKeepaliveSeconds": 30
          }
        }
      ]
    }
  ]
}
```

rastislavs replied:
awesome!

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -103,6 +107,14 @@ type BGPRouterManager struct {
+	// state management
+	StateReconcilers     []reconcilerv2.StateReconciler
+	StateNotifications   map[string]types.StateNotificationCh
+	PendingInstances     sets.Set[string]
```
all other fields have `State` in the name, maybe `StatePendingInstances`? Or what about wrapping all of these under a State struct?

rastislavs commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -103,6 +107,14 @@ type BGPRouterManager struct {
+	PendingInstances     sets.Set[string]
+	StateMutex           lock.Mutex // protects PendingInstances set.
+	SignalStateReconcile chan struct{}
+	SignalStateDeletion  chan string
```
On the first read, I was wondering what "state deletion" is and why it is of type string. I think it should have "instance deletion" in the name, + I would add a comment that the instance name is signalled.

rastislavs reviewed (approved):
Awesome work!

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -0,0 +1,277 @@
+		r.Logger.Debug("Update job running")
+
+		ticker := time.NewTicker(CRDStatusUpdateInterval)
+		defer ticker.Stop()
```
The purpose of rate-limiting the state reconciliation is to protect the API server. Then, we should do exponential backoff on failure and make a random jitter for every interval to avoid synchronized access. We should also make this interval configurable. Otherwise, users have nothing to do when they notice that BGP CPlane is overwhelming their API server.

harsimran-pabla replied:
Added exponential backoff on failure. I'd like to discuss more about configurable interval for updating CRD status. Should it be configurable via CRD or via cilium configmap ?

YutaroHayakawa replied:
I vote for the configMap option. I think this interval is about how much delay users can accept for status report and I believe this value rarely changes.

On the other hand, this is also a trade off between API server load and propagation delay. So, users may want to change this value when their API server is overloaded. However, because changing the value requires BGP session reset (without GR), the operational cost is significant, so we should try to not overload the API server as much as possible even if users set a short interval.

I think it's just about following the best practices like:
- Add jitter to the interval to avoid synchronized write from multiple nodes
- ✅ Exponentially backoff when the API call fails
- ✅ Avoid API call whenever possible (e.g. no diff)

harsimran-pabla replied:
Added this [commit](https://github.com/cilium/cilium/pull/33411/commits/ced206e433c10eb5e1f2e20f63fe4b9ef67a8477) to introduce ticker with jitter. Using external dependency for tickerWithJitter implementation.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -0,0 +1,277 @@
+			case <-ticker.C:
+				err := r.reconcileCRDStatus(ctx)
+				if err != nil {
+					r.Logger.WithError(err).Error("Failed to update CRD status")
```
The error level may be too high here. When the network between the API server and the node is down, this returns an error, but it's expected for example.

harsimran-pabla replied:
Added backoff: An error in the backoff is logged as debug and after backoff retries are exhausted, an error will be logged. I think it is reasonable not to flood the log messages but, at the same time, add an error log when the backoff count is reached.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -0,0 +1,277 @@
+
+func (r *StatusReconciler) Name() string {
+	return "CRDStatus"
+}
```
Maybe we should be more clear about which CRD status we're updating. Maybe `NodeConfigStatus`?

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -0,0 +1,277 @@
+		}
+
+		for _, runningPeerState := range peers.Peers {
+			if runningPeerState.PeerAddress == *configuredPeers.PeerAddress && runningPeerState.PeerAsn == *configuredPeers.PeerASN {
```
I'd change this to

```go
if runningPeerState.PeerAddress != *configuredPeers.PeerAddress || runningPeerState.PeerAsn != *configuredPeers.PeerASN {
    continue
}
```

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -0,0 +1,277 @@
+	client := r.ClientSet.CiliumV2alpha1().CiliumBGPNodeConfigs()
+	_, err = client.Patch(ctx, r.nodeName,
+		k8s_types.JSONPatchType, createStatusPatch, metav1.PatchOptions{
+			FieldManager: "cilium-bgp-agent",
```
I think this should align with the name of the Module.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -0,0 +1,277 @@
+	// do not reconcile if not in BGPv2 mode
+	if params.ConfigMode.Get() != mode.BGPv2 {
+		return nil
+	}
```
The state reconciliation won't happen in v1 right? Then we don't need this check?

https://github.com/cilium/cilium/pull/33411/commits/2e7392ba555abf80493e5a76a8324460f5ce9aed#diff-510521ec70b8612f036d50c65b9197d8737b661a561849cd4a02beb73b95c70fR144

harsimran-pabla replied:
Yes, reconciliation is not triggered from the source in the v1 scenario. I didn't want to rely on the underlying implementation for this check, as we may modify it in the future and cause issues in this reconciler. 

State reconcilers are all initialized at startup time, however v1 or v2 mode is dynamic and can change during runtime, so I wanted a way to propagate mode information to reconcilers and they can decide the behaviour based on the mode.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/crd_status.go:
```diff
@@ -0,0 +1,277 @@
+	current := &v2alpha1.CiliumBGPNodeStatus{}
+	if r.desiredStatus != nil {
+		current = r.desiredStatus.DeepCopy()
+	}
```
We can simply initialize r.desiredStatus with `&v2alpha1.CiliumBGPNodeStatus{}` in the constructor and change this to `current = r.desiredStatus.DeepCopy()`.

YutaroHayakawa commented on pkg/bgpv1/manager/manager.go:
```diff
@@ -47,10 +49,43 @@ type LocalInstanceMap map[string]*instance.BGPInstance
+	ConfigMode       *mode.ConfigMode
+	Reconcilers      []reconciler.ConfigReconciler   `group:"bgp-config-reconciler"`
+	ReconcilersV2    []reconcilerv2.ConfigReconciler `group:"bgp-config-reconciler-v2"`
+	StateReconcilers []reconcilerv2.StateReconciler  `group:"bgp-state-reconciler"`
```
bgp-state-reconciler => bgp-state-reconciler-v2 for consistency? This also tells readers that the state reconciliation is only for BGPv2.

YutaroHayakawa reviewed (changes requested):
Made an initial review. Nice stuff!

harsimran-pabla committed:
bgpv2: add jitter to status updates

Adding ticker with jitter in BGP CRD status updates. BGP changes are
generally done simultaneously on all nodes, which will result in status
update of CiliumBGPNodeConfig resource. Since there can be lot of nodes
in the cluster, we want to avoid all nodes updating status at the same
time.

This commit also introduces external dependency github.com/lthibault/jitterbug
which implements ticker with jitter.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

tommyp1ckles commented on pkg/k8s/apis/cilium.io/v2alpha1/bgp_node_types.go:
```diff
@@ -188,20 +189,16 @@ type CiliumBGPNodePeerStatus struct {
 	//
 	// +kubebuilder:validation:Optional
-	Uptime *string `json:"uptime,omitempty"`
+	EstablishedTime *string `json:"establishedTime,omitempty"`
```
Are there any considerations with respect to cilium upgrades here?

harsimran-pabla replied:
The CiliumBGPNodeConfig resource is internal; users do not modify it. Status fields were never set, so there are no consumers. I believe changing the fields should be safe. 

I remember I needed to bump the CRD version number for the upgrade so that it would automatically pick the new CRD version and update it in the latest push.

tommyp1ckles commented on pkg/bgpv1/manager/state_tracker.go:
```diff
@@ -0,0 +1,108 @@
+	"github.com/cilium/cilium/pkg/bgpv1/types"
+)
+
+// trackInstanceStateChange is a goroutine that listens on the tracker channel and signals state reconciler.
```
to clarify a bit, this is a function that is to be run as a goroutine, trackInstanceStateChanges does not create its own goroutine.

harsimran-pabla replied:
Yes, this function is run as a go-routine; I fixed the comment to clarify it.

tommyp1ckles commented on pkg/bgpv1/manager/state_tracker.go:
```diff
@@ -0,0 +1,108 @@
+		Debug("Reconciling state for instances with pending state changes")
+
+	// process all pending instances
+	for instanceName := range m.state.pendingInstances {
```
I'm not that familiar with the bgpv2 code, however this seems like it may lack some resiliency? That is, `reconcileInstanceState(...)` can fail and will log the error, but regardless of that outcome m.state.pendingInstances is cleared. 

Does that mean that the state update is lost in this case?  Would it make sense to keep those instances in the set if the reconcile fails?

harsimran-pabla replied:
Thanks for this feedback; I modified the code to bubble up the error from the reconciliation error. On failure, we will retry with some backoff.

harsimran-pabla replied:
State updates are triggers. The reconciler itself checks the actual state change by calling the underlying gobgp instance. Losing an event is not an issue as long as another event is fired.

tommyp1ckles commented on pkg/bgpv1/manager/state_tracker.go:
```diff
@@ -0,0 +1,108 @@
+		}
+	}
+
+	// tracker is close, signal the main reconcile loop that this instance is deleted so it
```
typo: tracker is ~close~ -> closed

tommyp1ckles commented on pkg/bgpv1/manager/state_tracker.go:
```diff
@@ -0,0 +1,108 @@
+		// It is okay if that channel is full, there can be multiple instances trying to
+		// notify the main reconcile loop that the state has changed.
+		select {
+		case m.state.reconcileSignal <- struct{}{}:
```
Just thinking out loud: this feels a bit like the statedb table watch mechanism, where we watch for updates on a set of data entries.

harsimran-pabla replied:
The current approach mostly signals to reconcilers when a state change is detected. We are not pushing any actual data around, it was not clear to me if statedb would be appropriate for this mechanism.

From some reconcilers we may need to propagate routing information out to other subsystems, statedb would make lot of sense for that solution.

tommyp1ckles reviewed (changes requested):
nice work @harsimran-pabla 😄  

K8s Changes look good, but I did have some comments regarding the reconciler code

tommyp1ckles commented on pkg/bgpv1/manager/state_tracker.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// reconcileState is the main loop that reconciles the state of all instances that have pending state changes.
+// It will take StateMutex lock to process the pending instances and then reset the pending instances set.
```
nit: `then reset the pending instances set.`-> `and remove them from the pending instances set`

YutaroHayakawa reviewed (approved):
> Intruducing cancel context when starting go bgp instance. We need to close this context so watch goroutines can exit gracefully.

This sounds like a GoBGP's bug. Does this mean we were leaking the Watcher Goroutine? At a glance, GoBGP doesn't use the context passed to the `StartBgp`.

learnitall reviewed (approved):
Vendor changes LGTM, thanks!

rastislavs commented on pkg/bgpv1/gobgp/server.go:
```diff
@@ -81,6 +82,10 @@ type GoBGPServer struct {
 	server *server.BgpServer
+
+	// stopping is a flag to indicate if the server is stopping.
+	stopping  bool
+	stopMutex lock.Mutex
```
nit: if the mutex if guarding just this single bool value, maybe we could use atomic Bool instead?

harsimran-pabla replied:
I want to protect a critical section with the lock. In the watchers, checking the stopping flag and writing to the state update channel need to be protected. 

With atomic bool, we potentially have a race where an inflight route update can cause a write to the closed channel.

rastislavs replied:
got it, ack
